aana/processors/speaker.py

```diff
@@ -0,0 +1,633 @@
+        )
+        return best_match["speaker"]
+
+
+def get_first_word_idx_of_sentence(
+    word_idx: int,
+    word_list: list[str],
+    speaker_list: list[str | None],
+    max_words: int,
+) -> int:
+    """Get the index of the first word of the sentence in the given range."""
+    left_idx = word_idx
+    while (
+        left_idx > 0
+        and word_idx - left_idx < max_words
+        and speaker_list[left_idx - 1] == speaker_list[left_idx]
+        and word_list[left_idx - 1][-1] not in sentence_ending_punctuations
+    ):
+        left_idx -= 1
+
+    return (
+        left_idx
+        if left_idx == 0 or word_list[left_idx - 1][-1] in sentence_ending_punctuations
+        else -1
+    )
+
+
+def get_last_word_idx_of_sentence(
+    word_idx: int, word_list: list[str], max_words: int
+) -> int:
+    """Get the index of the last word of the sentence in the given range."""
+    right_idx = word_idx
+    while (
+        right_idx < len(word_list)
+        and right_idx - word_idx < max_words
+        and word_list[right_idx][-1] not in sentence_ending_punctuations
+    ):
+        right_idx += 1
+
+    return (
+        right_idx
+        if right_idx == len(word_list) - 1
+        or word_list[right_idx][-1] in sentence_ending_punctuations
+        else -1
+    )
+
+
+def find_nearest_speaker(
+    index: int, word_speaker_mapping: list[AsrWord], reverse: bool = False
+) -> str | None:
+    """Find the nearest speaker label in the word_speaker_mapping either forward or backward.
+
+    Args:
+        index (int): The index to start searching from.
+        word_speaker_mapping (list[AsrWord]): List of word-speaker mappings.
+        reverse (bool): Search backwards if True; forwards if False. Default is False.
+
+    Returns:
+        str | None: The nearest speaker found or None if not found.
+    """
+    step = -1 if reverse else 1
+    for i in range(index, len(word_speaker_mapping) if not reverse else -1, step):
+        if word_speaker_mapping[i].speaker:
+            return word_speaker_mapping[i].speaker
+    return None
+
+
+def align_with_punctuation(
+    transcription: WhisperOutput, max_words_in_sentence: int = 50
+) -> list[AsrWord]:
+    """Aligns speaker labels with sentence boundaries defined by punctuation.
+
+    Args:
+        transcription (WhisperOutput): transcription with speaker information.
+        max_words_in_sentence (int): Maximum number of words allowed in a sentence.
+
+    Returns:
+        word_speaker_mapping: (list[AsrWord]): Realigned word-speaker mappings.
+    """
+    new_segments = [segment.words for segment in transcription["segments"]]
+    word_speaker_mapping = [word for segment in new_segments for word in segment]
+    words_list = [item.word for item in word_speaker_mapping]
+    speaker_list = [item.speaker for item in word_speaker_mapping]
+
+    # Fill missing speaker labels by finding the nearest speaker
+    for i, item in enumerate(word_speaker_mapping):
```

Comment:
Is it the same as fill_nearest in get_speaker_for_interval?

Reply:
No, both are different. This one finds the speaker for the next word(s) if the current word does not have a speaker assigned to it. `get_speaker_for_interval` assigns each word/segment a corresponding speaker if there is max overlap between the word and the diarization segment. If fill_nearest is True, it assigns each word a corresponding speaker even if there is zero overlap.

Basically, in the first case, `the near to a word` is concerning neighboring words. In the latter,` near to a word/segment` means the overlap of diarization segment with the word, and if it doesn't exist, look for the nearest diarization segment (has nothing to do with a nearby word).